src/bbd/geocoder/geocoder.py

```diff
@@ -0,0 +1,582 @@
+    "456 Test Way"
+    >>> street_encode("420 Test Ave Unit 5")
+    "420 Test Ave"
+    """
+    pattern = r"^.*\S(?=(?:\s+)(?:[Aa]p(?:ar)?t(?:ment)? |[Uu]nit |#|[Ll]ot |"\
+              r"[Rr](?:oo)?m |[Bb](?:ui)?ld(?:in)?g |[Uu]pp?e?r|[Ll]o?we?r|"\
+              r"[Ss](?:ui)?te |[Ff](?:l )?(?:rnt)?|[Tt]r(?:ai)?le?r |[Dd]ept ))"
+    try:
+        result = re.search(pattern, street) or street
+    except TypeError:
+        result = ""
+    if type(result) == str:
+        return result #No match
+    else:
+        return result[0] #Match
+    #try: #Python >=3.7
+    #    if type(result) is re.Match:
+    #        #want only first result.
+    #        return result[0]
+    #    return result
+    #except AttributeError: #Python<3.7
+    #    if type(result) == str:
+    #        return result
+    #    else:
+    #        return result[0]
+
+
+class GeocodeLocations:
+    """ Friendly wrapper to geocode large batches of data with Nominatim
+    and saves to file in tab deliminated format located at path. This 
+    will remember your progress across session and also allows user to 
+    only run a number of batches of size batch_size.
+
+    Please note, this process MAY NOT be multi-threaded or run in parallel.
+    Please read Nominatim's Usage Policy here: 
+    https://operations.osmfoundation.org/policies/nominatim/
+
+    Attributes
+    ----------
+    data : pandas dataframe, list of dicts, or list of strings.
+        Data where each row or item represents 1 address to be geocoded. 
+        NOTE: Inputting falsey value for data allows for providing
+            column of components by kwarg as either Series or list of
+            str and setting defaults as necessary. (see defaults)
+        For dataframe: 
+            data should either contain 1 column
+            OR 
+            columns using follow VAN default column names (not case sensitive):
+             "Address" - column of str street component of address
+             "City" - column of str city component of address
+             "County" - column of str county component of address
+             "State" - column of str state component of address
+             "Zip5" - column of str 5-digit postal code for address
+             "Country" - column of str country component of address
+                         (if applicable, default will set country to US)
+        For list of dicts:
+            Each item should be a dict of the form
+             "Address" : str street component of address
+             "City" : str city component of address
+             "County" : str county component of address
+             "State" : str state component of address
+             "Zip5" : str 5-digit postal code for address
+             "Country" : str country component address
+                         (if applicable, default will set country to US)
+        For list of strings:
+            Each item should str full address to pass as query.
```

Comment:
So basically there are 2 ways the data can be formatted: 
1) it will be broken into components like 
```python
{
    "street":"101 test ave", 
    "city": "testville", 
    "county": "testis county",
    "state": "te~x~stas",
    "zip": "55555"
}
```
This is the standard format you would have with a dataframe loaded from VAN. Which is helpful since secondary unit designators (i.e Unit 1, Apartment 4, etc) will cause Nominatim to return null for an otherwise perfectly valid address. So I trim the street component w/ street_encode.

2) A string address e.g. "101 test ave. unit 2, testville, testis county, te~x~stas 55555"

Reply:
That makes sense will update